orttraining/orttraining/models/runner/training_util.cc

```diff
@@ -72,6 +72,39 @@ size_t DataSet::TotalBatch(size_t batch_size) const {
   return NumSamples() / batch_size + ((NumSamples() % batch_size > 0) ? 1 : 0);
 }
 
+// gather additional training params from tensor dimensions
+// see metrics_map in bert/main.cc for example, and training_utils.h for more explanation
+void DataSet::GetTensorDimensionsFromInputs(size_t batch_size, const std::map<std::string, std::pair<std::string, size_t>>& metrics_map,
+                                            MapStringToString& perf_properties) const {
+  if (metrics_map.size() == 0) return; 
+
+  batch_size = min(batch_size, data_.size());
+  for (size_t input_index = 0; input_index < NumInputs(); ++input_index) {
+    std::string input_name = GetInputName(input_index);
+    const auto it = metrics_map.find(input_name);
+    if (it == metrics_map.end()) continue;
+    auto metric = it->second;
+
+    const Tensor& first_tensor = data_[0]->at(input_index).Get<Tensor>();
+    std::vector<int64_t> shape_vector = first_tensor.Shape().GetDims();
+    // By default, batch_size(i.e., batch) will be at postition 0.
+    // although batch size normally is not used, we are not skipping it to make it look consistent with 
+    // the graph loaded by, say, Netron. Thus we are expecting training param (other than batch size) 
+    // would start from position 1. For example int64[batch,sequence] for input "input1" in bert model.
+    if (first_tensor.Shape().Size() > 1) {
+      shape_vector.insert(shape_vector.begin(), batch_size);
+    } else {
+      shape_vector.clear();
+      shape_vector.emplace_back(batch_size);
+    }
+    // skipping out of bounds position, continue with warning
+    if (metric.second > static_cast<size_t>(-1) && metric.second < shape_vector.size())
```

Comment:
i believe `static_cast<size_t>(-1)` is the max size_t value, so it looks like this condition is always false